File: bench/definitive/java/BinaryTrees.java
public class BinaryTrees {
    static class Node {
        int item;
        Node left, right;
        Node(int i) { this.item = i; }
        Node(int i, Node l, Node r) { this.item = i; this.left = l; this.right = r; }
    }

    static int itemCheck(Node n) {
        if (n.left == null) return n.item;
        return n.item + itemCheck(n.left) - itemCheck(n.right);
    }

    static Node bottomUpTree(int i, int d) {
        if (d > 0) return new Node(i, bottomUpTree(2 * i - 1, d - 1), bottomUpTree(2 * i, d - 1));
        return new Node(i);
    }

    public static void main(String[] args) {
        int minD = 4, maxD = 14, sD = maxD + 1;
        System.out.printf("Stretching tree of depth %d%n", sD);
        Node st = bottomUpTree(0, sD);
        int ch = itemCheck(st);
        System.out.printf("Item check: %d%n", ch);

        for (int depth = minD; depth <= maxD; depth += 2) {
            int it = 1 << (maxD - depth + minD);
            int sum = 0;
            for (int i = 1; i <= it; i++) {
                Node t1 = bottomUpTree(i, depth);
                Node t2 = bottomUpTree(-i, depth);
                sum += itemCheck(t1) - itemCheck(t2);
            }
            System.out.printf("%d trees of depth %d item check: %d%n", it * 2, depth, sum);
        }

        Node ll = bottomUpTree(0, maxD);
        System.out.printf("Long lived tree of depth %d item check: %d%n", maxD, itemCheck(ll));
    }
}

File: examples/benchmarks/binary-trees.cpp
// binary-trees.cpp — CLBG "binary-trees" benchmark em C++ (malloc/free manual)
// Equivalente direto ao binary-trees.hphl (mesma lógica, sem GC).

#include <cstdio>
#include <cstdlib>
#include <cstdint>

class Node {
public:
    int item;
    Node* left;
    Node* right;
    Node(int item) : item(item), left(nullptr), right(nullptr) {}
    Node(int item, Node* left, Node* right) : item(item), left(left), right(right) {}
    ~Node() {
        // destrutor recursivo: libera a subárvore inteira
        // (em C++ com std::unique_ptr isso seria automático; aqui é manual)
    }
};

int itemCheck(Node* node) {
    if (node->left == nullptr) {
        return node->item;
    }
    return node->item + itemCheck(node->left) - itemCheck(node->right);
}

// bottomUpTree aloca e retorna; caller é responsável por free
Node* bottomUpTree(int item, int depth) {
    if (depth > 0) {
        return new Node(item,
                        bottomUpTree(2 * item - 1, depth - 1),
                        bottomUpTree(2 * item, depth - 1));
    }
    return new Node(item, nullptr, nullptr);
}

// Libera a árvore inteira recursivamente
void freeTree(Node* node) {
    if (node == nullptr) return;
    freeTree(node->left);
    freeTree(node->right);
    delete node;
}

int main() {
    int minDepth = 4;
    int maxDepth = 14;
    int stretchDepth = maxDepth + 1;

    std::printf("Stretching tree of depth %d\n", stretchDepth);
    Node* stretchTree = bottomUpTree(0, stretchDepth);
    int check = itemCheck(stretchTree);
    freeTree(stretchTree);

    std::printf("Item check: %d\n", check);

    int longLivedTree = 0;
    for (int depth = minDepth; depth <= maxDepth; depth += 2) {
        int iterations = 1 << (maxDepth - depth + minDepth);
        int sum = 0;
        for (int i = 1; i <= iterations; i++) {
            Node* t1 = bottomUpTree(i, depth);
            Node* t2 = bottomUpTree(-i, depth);
            sum = sum + itemCheck(t1);
            sum = sum - itemCheck(t2);
            freeTree(t1);
            freeTree(t2);
        }
        std::printf("%d trees of depth %d item check: %d\n", iterations * 2, depth, sum);
    }

    Node* longLived = bottomUpTree(0, maxDepth);
    std::printf("Long lived tree of depth %d item check: %d\n", maxDepth, itemCheck(longLived));
    freeTree(longLived);
    return 0;
}
